packages/dss-bridge/src/DomainHost.sol

```diff
@@ -0,0 +1,305 @@
+
+            grain += minted;
+        }
+
+        line = rad;
+
+        _lift(rad, minted);
+
+        emit Lift(wad);
+    }
+
+    /// @notice Withdraw pre-mint DAI from the remote domain
+    function release(uint256 wad) external guestOnly vatLive {
+        int256 amt = -_int256(wad);
+
+        require(dai.transferFrom(escrow, address(this), wad), "DomainHost/transfer-failed");
+        daiJoin.join(address(this), wad);
+        vat.frob(ilk, address(this), address(this), address(this), amt, amt);
+        vat.slip(ilk, address(this), amt);
+
+        grain -= wad;
+
+        emit Release(wad);
+    }
+
+    /// @notice Merge DAI into surplus
+    function surplus(uint256 wad) external guestOnly {
+        dai.transferFrom(address(escrow), address(this), wad);
+        daiJoin.join(address(vow), wad);
+
+        emit Surplus(wad);
+    }
+
+    /// @notice Cover the remote domain's deficit by pulling debt from the surplus buffer
+    function deficit(uint256 wad) external guestOnly vatLive {
+        vat.suck(vow, address(this), wad * RAY);
+        daiJoin.exit(address(escrow), wad);
+        
+        // Send ERC20 DAI to the remote DomainGuest
+        _rectify(wad);
+
+        emit Deficit(wad);
+    }
+
+    /// @notice Initiate shutdown for this domain
+    /// @dev This will trigger the end module on the remote domain
+    function cage() external {
+        require(vat.live() == 0 || wards[msg.sender] == 1, "DomainHost/not-authorized");
+        require(live == 1, "DomainHost/not-live");
+
+        live = 0;
+
+        _cage();
+
+        emit Cage();
+    }
+
+    /// @notice Set this domain's cure value
+    function tell(uint256 value) external guestOnly {
+        require(live == 0, "DomainHost/live");
+        require(!cureReported, "DomainHost/cure-reported");
+        require(_divup(value, RAY) <= grain, "DomainHost/cure-bad-value");
+
+        cureReported = true;
+        cure = value;
+
+        emit Tell(value);
+    }
+
+    /// @notice Allow DAI holders to exit during global settlement
+    /// @dev    This will mint a pro-rata claim token on the remote domain.
+    ///         Gem amount is scaled by the actual debt of the remote domain.
+    ///         `usr` is the address for the mint on the remote domain.
+    function exit(address usr, uint256 wad) external {
+        require(vat.live() == 0, "DomainHost/vat-live");
+        vat.slip(ilk, msg.sender, -_int256(wad));
+
+        // Convert to actual debt amount
+        // Round against the user
+        uint256 claim = wad * (grain - _divup(cure, RAY)) / grain;
+        
+        _mintClaim(usr, claim);
+
+        emit Exit(usr, wad, claim);
+    }
+
+    // --- Canonical DAI Support ---
+
+    /// @notice Deposit local DAI to mint remote canonical DAI
+    /// @param to The address to send the DAI to on the remote domain
+    /// @param amount The amount of DAI to deposit [WAD]
+    function deposit(address to, uint256 amount) external {
```

Comment:
We would probably need to add `bytes calldata data` as extra argument to be able to include rollup-specific parameters (e.g. maxGas and gasPriceBid on arbitrum). And similary for `DomainGuest.withdraw()`.

Reply:
So this is what `_deposit()` would take care of. I wanted to keep all chain-specific stuff outside of this if possible. `maxGas` seems like it could be hard coded at the function level. `gasPriceBid` seems a bit more tricky.

Reply:
`maxGas` could be hardcoded but `gasPriceBid` needs to be estimated off-chain before each L2 call, so it would be required by the method.

Reply:
That's unfortunate. We may need to include arbitrary bytes on all functions. =\

Reply:
Reading the Arbitrum docs it appears as though the `gasPriceBid` is not necessary to create the ticket. You can execute the tx yourself later on. Whether this is a good idea or not we should discuss.

Reply:
Yeah I guess it's possible at the cost of a worse UX for depositors. But then there is also `maxSubmissionCost`, which can't be omitted and changes every 24 hours. But I guess the broader point here is that some rollups may have idiosyncrasies that require custom params to be included so I would prefer to include this extra `data` arg to be on the safe side.

Reply:
I have an idea to not clutter the function definitions too much if custom params are not needed like in the case of Optimism. I'm going to set all functions that need to send xchain message to `public` such that overloads can be defined by the inheriting contract. These functions will then stash the extra parameters in a memory slot which can be retrieved by the callback function when invoked. This will keep the parent contract clean, but allow any number of custom params to be invoked with well defined ABIs. So for example in ArbitrumDomainHost.sol:

```
function lift(uint256 wad, ... extra gasPrice, gasLimit, etc params) external {
	// Stash all the extra parameters in memory

	lift(wad);
}

function _lift(uint256 id, uint256 line, uint256 minted) internal override {
	// Load extra params from memory

	// Execute send message operation
}
```

Reply:
That's quite clever and based on some quick benchmarks this approach would save ~600 gas vs decoding the extra params from a generic `bytes calldata params`. But I'm a bit concerned by the loss of readability given the need for low level mstore/mload and the fact that the code is now split in at least 3 functions, spread across two files.

Reply:
Here is an example: https://github.com/makerdao/xdomain/pull/44/files#diff-4a71f917be7f3439a64c459dd82d1459e89b7971bae72d7fa71f8c47fd0eb709R57

I like that this allows default to not require `bytes` being added to the end of some of the functions, but granted it's a little bit of a hacky workaround. Would like to have discussion about this before proceeding.